pkg/agent/conversation.go

```diff
@@ -119,275 +208,547 @@ func (s *Conversation) Init(ctx context.Context, doc *ui.Document) error {
-	currChatContent = []any{query}
-
-	currentIteration := 0
-	maxIterations := a.MaxIterations
-
-	for currentIteration < maxIterations {
-		log.Info("Starting iteration", "iteration", currentIteration)
-
-		a.Recorder.Write(ctx, &journal.Event{
-			Timestamp: time.Now(),
-			Action:    "llm-chat",
-			Payload:   []any{currChatContent},
-		})
 
-		var agentTextBlock *ui.AgentTextBlock
-
-		// We create the agent text block here; this lets renderers render a "thinking" state
-		// before the first response arrives.
-		agentTextBlock = ui.NewAgentTextBlock()
-		agentTextBlock.SetStreaming(true)
-		a.doc.AddBlock(agentTextBlock)
-
-		stream, err := a.llmChat.SendStreaming(ctx, currChatContent...)
-		if err != nil {
-			return err
-		}
-
-		// Clear our "response" now that we sent the last response
-		currChatContent = nil
-
-		if a.EnableToolUseShim {
-			// convert the candidate response into a gollm.ChatResponse
-			stream, err = candidateToShimCandidate(stream)
+	log.Info("Starting agent loop", "initialQuery", initialQuery, "runOnce", c.RunOnce)
+	go func() {
+		if initialQuery != "" {
+			c.addMessage(api.MessageSourceUser, api.MessageTypeText, initialQuery)
+			answer, err := c.handleMetaQuery(ctx, initialQuery)
 			if err != nil {
-				return err
+				log.Error(err, "error handling meta query")
+				c.setAgentState(api.AgentStateDone)
+				c.pendingFunctionCalls = []ToolCallAnalysis{}
+				c.addMessage(api.MessageSourceAgent, api.MessageTypeError, "Error: "+err.Error())
+			} else if answer != "" {
+				// we handled the meta query, so we don't need to run the agentic loop
+				c.setAgentState(api.AgentStateDone)
+				c.pendingFunctionCalls = []ToolCallAnalysis{}
+				c.addMessage(api.MessageSourceAgent, api.MessageTypeText, answer)
+			} else {
+				// Start the agentic loop with the initial query
+				c.setAgentState(api.AgentStateRunning)
+				c.currIteration = 0
+				c.currChatContent = []any{initialQuery}
+				c.pendingFunctionCalls = []ToolCallAnalysis{}
 			}
+		} else {
+			greetingMessage := "Hey there, what can I help you with today ?"
+			c.addMessage(api.MessageSourceAgent, api.MessageTypeUserInputRequest, greetingMessage)
 		}
+		for {
+			var userInput any
+			log.Info("Agent loop iteration", "state", c.AgentState())
+			switch c.AgentState() {
+			case api.AgentStateIdle, api.AgentStateDone:
+				// In RunOnce mode, we are done, so exit
+				if c.RunOnce {
+					log.Info("RunOnce mode, exiting agent loop")
+					c.setAgentState(api.AgentStateExited)
+					return
+				}
+				log.Info("initiating user input")
+				c.addMessage(api.MessageSourceAgent, api.MessageTypeUserInputRequest, ">>>")
+				select {
+				case <-ctx.Done():
+					log.Info("Agent loop done")
+					return
+				case userInput = <-c.Input:
+					log.Info("Received input from channel", "userInput", userInput)
+					if userInput == io.EOF {
+						log.Info("Agent loop done, EOF received")
+						return
+					}
+					query, ok := userInput.(*api.UserInputResponse)
```

Comment:
Should we have more scrutiny in input validation?

Reply:
We should probably but not clear how and what to validate. we need to sanitize all inputs to the agent. user input is still safe, the output from the tool calls is a major case for attack vector.